Main.py

```diff
@@ -149,45 +149,45 @@ def main(args, seed=None, baked_server_options: Optional[Dict[str, object]] = No
-        new_items: List[Item] = []
-        old_items: List[Item] = []
-        depletion_pool: Dict[int, Dict[str, int]] = {
-            player: getattr(multiworld.worlds[player].options,
-                            "start_inventory_from_pool",
-                            StartInventoryPool({})).value.copy()
-            for player in multiworld.player_ids
-        }
-        for player, items in depletion_pool.items():
-            player_world: AutoWorld.World = multiworld.worlds[player]
-            for count in items.values():
-                for _ in range(count):
-                    new_items.append(player_world.create_filler())
-        target: int = sum(sum(items.values()) for items in depletion_pool.values())
-        for i, item in enumerate(multiworld.itempool):
+    fallback_inventory = StartInventoryPool({})
+    depletion_pool: Dict[int, Dict[str, int]] = {
+        player: getattr(multiworld.worlds[player].options, "start_inventory_from_pool", fallback_inventory).value.copy()
+        for player in multiworld.player_ids
+    }
+    target_per_player = {player: sum(target_items.values()) for player, target_items in depletion_pool.items()}
+    overall_target = sum(target_per_player.values())
+
+    if overall_target:
+        new_itempool: List[Item] = []
+
+        # Make new itempool with start_inventory_from_pool items removed
+        for item in multiworld.itempool:
             if depletion_pool[item.player].get(item.name, 0):
-                target -= 1
                 depletion_pool[item.player][item.name] -= 1
-                # quick abort if we have found all items
-                if not target:
-                    old_items.extend(multiworld.itempool[i+1:])
+                overall_target -= 1
+                if overall_target == 0:
                     break
             else:
-                old_items.append(item)
-
-        # leftovers?
-        if target:
-            for player, remaining_items in depletion_pool.items():
-                remaining_items = {name: count for name, count in remaining_items.items() if count}
-                if remaining_items:
-                    logger.warning(f"{multiworld.get_player_name(player)}"
-                                    f" is trying to remove items from their pool that don't exist: {remaining_items}")
-                    # find all filler we generated for the current player and remove until it matches 
-                    removables = [item for item in new_items if item.player == player]
-                    for _ in range(sum(remaining_items.values())):
-                        new_items.remove(removables.pop())
-        assert len(multiworld.itempool) == len(new_items + old_items), "Item Pool amounts should not change."
-        multiworld.itempool[:] = new_items + old_items
+                new_itempool.append(item)
+
+        unfound_items_per_player = {
```

Comment:
I feel like this section could all be inside an `if overall_target:` and then at the end have:
```py
else:
    for player, target in target_per_player.items():
        new_itempool += [multiworld.worlds[player].create_filler() for _ in range(target_per_player[player])]
```

Reply:
I do think this makes the code slightly more complicated, at probably little performance gain (The full iteration of the multiworld itempool is gonna be so much slower than anything that happens after it)

But I see your point. I'll think about it. I don't like the idea of just wrapping it in an extra indent, but maybe there is a different cute solution here.

Reply:
I timed it roughly, this change barely has any impact.

Reply:
Yeah, I went and timed it myself with 10K players and it saved 0.0054 seconds on average over 20 trials (the range was 0.005-0.006) so... yeah, not worth

Reply:
I made a different improvement that should make this better in the worst cases